keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+- Unifies with the StorageClass API. Although ResourceClass shares many similar motivations and requirements as the existing StorageClass API, they focus on different kinds of resources. StorageClass is used to represent storage resources that are stateful and contains special storage semantics. ResourceClass, on the other hand, focuses on stateless compute resources, whose usage is bound to container lifecycle and can’t be shared across multiple nodes at the same time. For these reasons, we don’t plan to unify the two APIs.
+- Resource overcommitment, fractional resource requirements, native compute resource (i.e., cpu and memory) with special metadata requirements, and group compute resources. They are out of our current scope.
+
+## Components
+### ResourceClass API
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying node level compute resource plus metadata matching. We can extend this API to better support cluster resources and group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // defines general resource property matching constraints.
+        // e.g.: zone in { us-west1-b, us-west1-c }; type: k80
+        MetadataRequirements metav1.LabelSelector
+        // used to compare preference of two matching ResourceClasses
+        // The purpose to introduce this field is explained more later
+        Priority int
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "nvidia-gpu"
+        - key: "memory"
+          operator: "GtEq"
+          values:
+            - "30G"
+
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  containers:
+    - name: example-container
+      resources:
+        limits:
+          nvidia.high.mem: 2
```

Comment:
What if the ResourceClass name is the same as one non-native kubernetes raw resource name ?

Reply:
> What if the ResourceClass name is the same as one non-native kubernetes raw resource name?

The objectives section (migration path) suggests it will be possible for a resource class to override an existing resource. If the plan is to migrate native resources to use ComputeResource as well, then there may not be a need to discriminate between native and non-native resources in the scheduler. If the scheduler logic looks to match resources classes first before attempting a fit based on the 'legacy' allocatable resources, it may just work out. (But all this should be made explicit in the doc.)

cc @vikaschoudhary16 @jiayingz

Reply:
Compute Resources will not be advertised in Node.Status.Allocatables rather would be advertised in Node.Status.ComputeResources. 
> What if the ResourceClass name is the same as one non-native kubernetes raw resource name ?

In that case ResourceClass creation will be allowed only if it has no selectors. Also priority of such resource class will be set very high by default.  Requesting such a resource class in pod spec will be same effect as requesting raw compute resource name.

Reply:
@NickrenREN Your comment on putting resource quantity in ResourceClass sparked a thought.  Sometimes certain combinations of physical resources are themselves a unique resource.

Example: If I have a box with four NICs, N1, N2, N3, N4.  N1 and N3 may be connected to a TOR switch in such a way that they can be channel bonded *with each other*.  But not with the other NICs.  

Note: I am not suggesting ResourceClasses have to solve this kind of problem.  I just throw it out there as an interesting example of a class of things I suspect is going to be true in other places.